src/main/java/com/amazon/opendistroforelasticsearch/sql/parser/subquery/SubqueryRewriter.java

```diff
@@ -0,0 +1,30 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.parser.subquery;
+
+import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlSelectQueryBlock;
+
+/**
+ * Rewrite the Predicate Subquery to JOIN.
+ */
+public interface SubqueryRewriter {
+
+    /**
+     * Rewrite the subquery with correspond JOIN query.
+     * It is NOT idempotent operation.
+     */
+    void rewrite(MySqlSelectQueryBlock query);
```

Comment:
Why not keep the input parameter intact and return the re-written one as a result?

Reply:
I think it is not necessary. In this case, we need copy the MySqlSelectQueryBlock.